src/pipeline/src/etl.rs

```diff
@@ -97,23 +223,178 @@ impl<T> Pipeline<T>
 where
     T: Transformer,
 {
-    pub fn exec(&self, val: value::Value) -> Result<T::Output, String> {
-        let mut val = val;
+    fn exec_map<'a>(&self, map: &'a mut Map) -> Result<&'a mut Map, String> {
```

Comment:
this function already accepts aa mutable reference to map, so it doesn't have to return it. Just use `Result<(), String>` to make it clear.